core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java

```diff
@@ -67,93 +51,47 @@
-        this.clusterService = clusterService;
-        this.ingestService = ingestService;
-        this.ingestForwarder = new IngestActionForwarder(transportService);
-        clusterService.addStateApplier(this.ingestForwarder);
+                actionFilters, indexNameExpressionResolver, IndexRequest::new, IndexRequest::new, ThreadPool.Names.INDEX);
+        this.bulkAction = bulkAction;
+        this.shardBulkAction = shardBulkAction;
     }
 
     @Override
     protected void doExecute(Task task, final IndexRequest request, final ActionListener<IndexResponse> listener) {
-        if (Strings.hasText(request.getPipeline())) {
-            if (clusterService.localNode().isIngestNode()) {
-                processIngestIndexRequest(task, request, listener);
-            } else {
-                ingestForwarder.forwardIngestRequest(IndexAction.INSTANCE, request, listener);
-            }
-            return;
-        }
-        // if we don't have a master, we don't have metadata, that's fine, let it find a master using create index API
-        ClusterState state = clusterService.state();
-        if (shouldAutoCreate(request, state)) {
-            CreateIndexRequest createIndexRequest = new CreateIndexRequest();
-            createIndexRequest.index(request.index());
-            createIndexRequest.cause("auto(index api)");
-            createIndexRequest.masterNodeTimeout(request.timeout());
-            createIndexAction.execute(task, createIndexRequest, new ActionListener<CreateIndexResponse>() {
-                @Override
-                public void onResponse(CreateIndexResponse result) {
-                    innerExecute(task, request, listener);
-                }
-
-                @Override
-                public void onFailure(Exception e) {
-                    if (ExceptionsHelper.unwrapCause(e) instanceof ResourceAlreadyExistsException) {
-                        // we have the index, do it
-                        try {
-                            innerExecute(task, request, listener);
-                        } catch (Exception inner) {
-                            inner.addSuppressed(e);
-                            listener.onFailure(inner);
-                        }
-                    } else {
-                        listener.onFailure(e);
-                    }
+        BulkRequest bulkRequest = new BulkRequest();
+        bulkRequest.add(request);
+        bulkRequest.setRefreshPolicy(request.getRefreshPolicy());
+        bulkRequest.timeout(request.timeout());
+        bulkRequest.waitForActiveShards(request.waitForActiveShards());
+        request.setRefreshPolicy(WriteRequest.RefreshPolicy.NONE);
+        bulkAction.execute(task, bulkRequest, new ActionListener<BulkResponse>() {
+            @Override
+            public void onResponse(BulkResponse bulkItemResponses) {
+                assert bulkItemResponses.getItems().length == 1: "expected only one item in bulk request";
```

Comment:
since this is also the same code here as in delete, I wonder if we should add a generic BWC class that will do all this shared munging (and the code from the bulk action as well).